### PR TITLE
[XLA:GPU] Add a check to help prevent future bug

### PR DESCRIPTION
**Pull request by nouiz**
https://github.com/tensorflow/tensorflow/commit/67adc58cd64f3b86417d4466b6b656ecb8df3b2f
Fixed a wrong value bug. We got it by that for a long time and lost too much time due to it in parallel to you.

So to help prevent it from coming back, I added some extra check in the code. So if the codegen have this bug again, it will crash instead of producing wrong value.

@cheshire

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 1]
```diff
@@ -4373,18 +4373,21 @@ void IrEmitterUnnested::EmitPrologueForReduction(
 
 void IrEmitterUnnested::EmitFullWarpShuffleDownLoopForAllReduces(
     absl::Span<HloComputation* const> reducers,
-    absl::Span<llvm::AllocaInst* const> partial_result_addresses) {
+    absl::Span<llvm::AllocaInst* const> partial_result_addresses,
+    int block_size) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:4377]: Any reason to not say `thread_per_block` to be consistent with the `GetThreadsPerBlock`?

**Review by sanjoy — COMMENTED** [review]
Minor nit.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:4377]: I amended the commit.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -4373,18 +4373,21 @@ void IrEmitterUnnested::EmitPrologueForReduction(
 
 void IrEmitterUnnested::EmitFullWarpShuffleDownLoopForAllReduces(
     absl::Span<HloComputation* const> reducers,
-    absl::Span<llvm::AllocaInst* const> partial_result_addresses) {
+    absl::Span<llvm::AllocaInst* const> partial_result_addresses,
+    int thread_per_block) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:4377]: Sorry for nit-picking, but should be `threads_per_block`.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 2 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Amended.
I saw some old comments that had used the singular form. I fixed them in a new commit.

**Review by sanjoy — APPROVED** [review]

**Comment by nouiz** [comment]
The CI failure doesn't look related to my changes. It is an OOM error and my changes doesn't affect the memory used.

**Comment by cheshire** [comment]
let's try few more times, we've recently fixed an OOM in the OSS build

**Comment by byronyi** [comment]
@nouiz We also traced down a XLA/GPU bug with respect to dynamic scale with `tf.math.reduce_all` in `tf.keras.mixed_precision.LossScaleOptimizer` to check if the gradients overflow. 

The commit you mentioned 67adc58 somehow fixed this issue but we don't know why. 

This issue could be reproduced using official model garden RN50 with ``--loss_scale=dynamic``, under TF 2.3-2.4. TF 2.2 and nightlies after 67adc58 (tf-nightly==2.4.0.dev20201018) are not affected.

/cc @reedwm

**Comment by cheshire** [comment]
@byronyi Model garden models do not include XLA compilation though? Or are you using the CompiledTransformer variant?

**Comment by byronyi** [comment]
> @byronyi Model garden models do not include XLA compilation though? Or are you using the CompiledTransformer variant?

It's ResNet50 Keras model with XLA enabled. We use the r2.2.0 branch: https://github.com/tensorflow/models/tree/r2.2.0

And the following command benchmark case on V100:

```
PYTHONPATH=$PWD/official \
python3 official/vision/image_classification/resnet/resnet_imagenet_main.py \
    --dd=<imagenet_dataset> \
    --ds=mirrored --ara=nccl \
    --bs=3072 \
    --dtype=fp16 \
    --enable_xla \
    --use_tensor_lr \
    --datasets_num_private_threads=48 \
    --train_epochs=90 \
    --ebe=10 \
    --enable_checkpoint_and_export=True \
    --ls=dynamic   # only affect when dynamic loss scaling is enabled
```

This issue has affected us since TF 2.3.0, and only got fixed in nightly after 67adc58.

**Comment by byronyi** [comment]
Found the offending commit c03e49d2c2d309cfa0c41d0892d9ffed5c485c4e.

It seems `__shfl*` intrinsics produces invalid values when block dim < 32.

**Comment by nouiz** [comment]
I think __shfl* instruction can be used when the block dim < 32.
The problem is the code that call them suppose block dim >= 32.
The function name tell the limitation: `EmitFullWarpShuffleDownLoopForReduce`. A full warp is 32 threads.

I could have updated it work with small power of 2 or complex cases.
But using so few threads per block isn't good CUDA programming and will give very poor performance.
So instead of implementing a complex solution that shouldn't be used, I just make it raise an error. So no wrong value will appear in the future.

If we ever need to support cases when block dim < 32 in the future, we can just add it then.

**Comment by sanjoy** [comment]
Did we resolve the issue?  Should https://github.com/tensorflow/tensorflow/commit/c03e49d2c2d309cfa0c41d0892d9ffed5c485c4e be rolled back?

CC @akuegel

**Comment by byronyi** [comment]
> Did we resolve the issue? Should [c03e49d](https://github.com/tensorflow/tensorflow/commit/c03e49d2c2d309cfa0c41d0892d9ffed5c485c4e) be rolled back?
> 
> CC @akuegel

It's been fixed by 67adc58
